borg/archiver.py

```diff
@@ -353,6 +355,155 @@ def do_extract(self, args):
+            if count is None:
+                return "<deleted>"
+            return format_file_size(count)
+
+        def compare_chunks(lhs, rhs, lhs_archive, rhs_archive):
+            lhs_hash = sha256()
+            rhs_hash = sha256()
+            for chunk in lhs_archive.pipeline.fetch_many(lhs):
+                lhs_hash.update(chunk)
+            for chunk in rhs_archive.pipeline.fetch_many(rhs):
+                rhs_hash.update(chunk)
+            return lhs_hash.digest() == rhs_hash.digest()
+
+        def compare_items(path, item, compare_item, deleted=False):
+            """
+            Compare two items with identical paths.
+            :param deleted: Whether one of the items has been deleted
+            """
+            if item[b'mode'] != compare_item[b'mode'] and not deleted:
+                if args.json:
+                    dump_json({
+                        "path": path,
+                        "what": "mode",
+                        "ref": stat.filemode(item[b'mode']),
+                        "cmp": stat.filemode(compare_item[b'mode'])
+                    })
+                else:
+                    print(remove_surrogates(path), "different mode")
+                    print("\t", args.location.archive, stat.filemode(item[b'mode']))
+                    print("\t", args.compare, stat.filemode(compare_item[b'mode']))
+                return
+            if not stat.S_ISREG(item[b'mode']) and not deleted:
+                return
+            if b'chunks' not in item or b'chunks' not in compare_item:
+                # This is a link
+                if item.get(b"source", None) != compare_item.get(b"source", None):
+                    if args.json:
+                        dump_json({
+                            "path": path,
+                            "what": "link",
+                            "ref": item.get(b"source", "<regular file>"),
+                            "cmp": compare_item.get(b"source", "<regular file>")
+                        })
+                    else:
+                        print(remove_surrogates(path), "different link")
+                        print("\t", args.location.archive, item.get(b'source', '<regular file>'))
+                        print("\t", args.compare, compare_item.get(b'source', '<regular file>'))
+                return
+            if item[b'chunks'] != compare_item[b'chunks'] or deleted or not can_compare_chunks:
+                # Contents are different
+                ref_chunks = {c[0] for c in item[b'chunks']}
+                cmp_chunks = {c[0] for c in compare_item[b'chunks']}
+                total_ref = None if not item[b'mode'] else sum(c[1] for c in item[b'chunks'])
+                total_compare = None if not compare_item[b'mode'] else sum(c[1] for c in compare_item[b'chunks'])
+                if (not can_compare_chunks and total_ref == total_compare and not deleted and
+                        compare_chunks(ref_chunks, cmp_chunks, ref, compare)):
```

Comment:
I'll rename that to `can_compare_chunk_ids`. And compare_chunks() to compare_chunk_contents()